Save all available voview metadata

JamesWrigley opened the pull request:
Previously only metadata for version 1.0 was written.

takluyver commented:
I think we should mark the output as dataFormatVersion 1.0 when the input files are 1.0, 1.1 or 1.2, because we don't want to deal with the differences between those. 1.1 changed the meaning of `INDEX/flag` (this was thankfully short lived), 1.2 changed it back and also added the `INDEX/origin` field, which EXtra-data currently doesn't use for anything.

takluyver commented on extra_data/tests/conftest.py:
```diff
@@ -8,7 +8,7 @@
 
 
-@pytest.fixture(scope='session', params=['0.5', '1.0'])
+@pytest.fixture(scope='session', params=['0.5', '1.0', '1.2'])
```
I'm not super keen on running a whole load of tests again with only a trivial difference, but we can live with that.

If we are going to mock format version 1.2, though, we should do it properly, which means adding the `INDEX/origin` dataset to the mocked data. I think it's fine to fill it all with `-1` for now.

https://github.com/European-XFEL/EXtra-data/blob/dc66ddab58d6d7d0a05a318c5fdb0c76a07e5114/extra_data/tests/mockdata/base.py#L189-L198

JamesWrigley replied:
Alrighty, added in 3cb8ded.

takluyver commented on extra_data/writer.py:
```diff
@@ -117,7 +117,8 @@ def write_metadata(self):
 
         metadata_grp = self.file.create_group('METADATA')
-        if metadata.get('dataFormatVersion') == '1.0':
+        format_version = float(metadata.get('dataFormatVersion'))
+        if format_version >= 1.0:
```
```suggestion
        if metadata.get('dataFormatVersion') != '0.5':
```

The float comparison would work here, but we should avoid converting version numbers to floats just as a general thing, because version 1.10 is different from 1.1.

philsmt replied:
The objects returned by `pkg_resources.parse_version` can be compared using inequality operators as expected for version numbers. I think we can assume `pkg_resources` to be available anyway with `setuptools`/`distutils` and it should not add a dependency.

takluyver replied:
`pkg_resources` can be very slow to import, though - it scans all the other installed packages when it's imported. So I strongly prefer to avoid using it if at all possible.

I always used to use a LooseVersion class from distutils for this, but now distutils is deprecated - though we could still use it for now - I think it will be gone in Python 3.12 or something. Or we could do a very simple bit of parsing ourselves, like `[int(m) for m in re.findall('\d+', version)] .

So I stick to `==` and `!=` whenever possible :wink:

philsmt replied:
TIL! Quick googling lead me to this: https://stackoverflow.com/a/11887885
Could that be an option? It seems to be installed in my `extra_data` dev environment without having done so manually, and I didn't notice any perceivable slowness when importing (unlike `pkg_resources` tested directly afterwards).

takluyver replied:
Yup, good point, `packaging` is an option too. Technically it's a new dependency, but it's one that's pretty widely available.

JamesWrigley replied:
Yeouch good catch, I did not realize `float()` had that behaviour :fearful: I decided to go with `packaging` because it's also useful in the tests for 1.2 (3cb8ded). Fixed in c4f1cd7.

JamesWrigley commented:
> I think we should mark the output as dataFormatVersion 1.0 when the input files are 1.0, 1.1 or 1.2, because we don't want to deal with the differences between those. 1.1 changed the meaning of `INDEX/flag` (this was thankfully short lived), 1.2 changed it back and also added the `INDEX/origin` field, which EXtra-data currently doesn't use for anything.

Roger that, added in c4f1cd7.

(and apologies for taking so long to get back to this, last couple of weeks have been hectic...)

tmichela commented:
LGTM

takluyver commented:
LGTM too, and sorry for taking way, way longer to get back to this. :speak_no_evil:
